src/main/cpp/src/timezones.cu

```diff
@@ -243,6 +243,246 @@ std::unique_ptr<column> convert_to_utc_with_multiple_timezones(
   return result;
 }
 
+// =================== ORC timezones begin ===================
+// ORC timezone uses java.util.TimeZone rules, which is different from java.time.ZoneId rules.
+
+/**
+ * @brief Get the transition index for the given time `time_ms` using binary search.
+ * The transition array is sorted, each int64 value is composed of 44 bits
+ * transition time in seconds and 20 bits offset in seconds.
+ * @param begin the beginning of the transition array.
+ * @param end the end of the transition array.
+ * @param time_ms the input time in milliseconds to find the transition index for.
+ */
+__device__ static int get_transition_index(int64_t const* begin,
```

Comment:
The following is a possibly invasive suggestion.  (If it doesn't fit per current timeline, we can consider it for a follow-up.)

The fact that we have explicit code for binary-search instead of using `thrust::lower_bound()` got me thinking:

Is there value in converting the transition array into milliseconds (after shifting it by 20) first?  Then the rest of the comparison might simply be a `lower_bound` call.  It will save shifting/multiplying log(numTransitions) times per input row.